core/src/main/java/com/linecorp/armeria/server/HttpStatusException.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2017 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import com.linecorp.armeria.common.Flags;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.util.Exceptions;
+
+/**
+ * A {@link RuntimeException} that is raised when an armeria internal http exception has occurred.
+ * This class is the general class of exceptions produced by a failed request or a reset stream.
+ */
+public class HttpStatusException extends RuntimeException {
+
+    private static final Map<Integer, HttpStatusException> EXCEPTIONS = new ConcurrentHashMap<>();
+
+    /**
+     * Returns a new {@link HttpStatusException} instance with the HTTP status code.
+     */
+    public static HttpStatusException of(int statusCode) {
+        return of(HttpStatus.valueOf(statusCode));
+    }
+
+    /**
+     * Returns a new {@link HttpStatusException} instance with the {@link HttpStatus}.
+     */
+    public static HttpStatusException of(HttpStatus httpStatus) {
+        requireNonNull(httpStatus, "httpStatus");
+        if (Flags.verboseExceptions()) {
+            return new HttpStatusException(httpStatus);
+        } else {
+            final int statusCode = httpStatus.code();
+            return EXCEPTIONS.computeIfAbsent(statusCode, code ->
+                    Exceptions.clearTrace(new HttpStatusException(code)));
+        }
+    }
+
+    private static final long serialVersionUID = 3341744805097308847L;
+
+    private final HttpStatus httpStatus;
+
+    /**
+     * Creates a new instance with HTTP status code.
+     */
+    public HttpStatusException(int statusCode) {
+        this(HttpStatus.valueOf(statusCode));
+    }
+
+    /**
+     * Creates a new instance.
+     */
+    public HttpStatusException(HttpStatus httpStatus) {
```

Comment:
Probably OK to make the public constructors private, because a user can use the static factory methods?